mirascope/integrations/logfire/_utils.py

```diff
@@ -0,0 +1,180 @@
+        "template_variables": result.fn_args,
+        "messages": result.messages,
+        "response_data": result.response,
+    }
+
+
+def get_stream_span_data(stream: BaseStream) -> dict:
+    content = ""
+    if stream.message_param:
+        content = stream.message_param.get("content", "") or stream.message_param.get(
+            "message", ""
+        )
+    return {
+        "messages": [stream.user_message_param],
+        "call_params": stream.call_params,
+        "model": stream.model,
+        "provider": stream._provider,
+        "prompt_template": stream.prompt_template,
+        "template_variables": stream.fn_args,
+        "output": {"cost": stream.cost, "content": content},
+    }
+
+
+def get_tool_calls(result: BaseCallResponse) -> list[dict] | None:
+    if tools := result.tools:
+        tool_calls = [
+            {
+                "function": {
+                    "arguments": tool.args,
+                    "name": tool._name(),
+                }
+            }
+            for tool in tools
+        ]
+        return tool_calls
+    return None
+
+
+def get_output(result: BaseCallResponse) -> dict[str, Any]:
+    output: dict[str, Any] = {}
+    if cost := result.cost:
+        output["cost"] = cost
+    if input_tokens := result.input_tokens:
+        output["input_tokens"] = input_tokens
+    if output_tokens := result.output_tokens:
+        output["output_tokens"] = output_tokens
+    if content := result.content:
+        output["content"] = content
+    return output
+
+
+def handle_call_response(
+    result: BaseCallResponse, logfire_span: logfire.LogfireSpan | None
+):
+    if logfire_span is None:
+        return
+    output = get_output(result)
+    span_data = get_call_response_span_data(result)
+    span_data["async"] = False
+    tool_calls = get_tool_calls(result)
+    if tool_calls:
+        output["tool_calls"] = tool_calls
+    span_data["output"] = output
+    logfire_span.set_attributes(span_data)
+
+
+def handle_stream(
+    stream: BaseStream,
+    logfire_span: logfire.LogfireSpan | None,
+):
+    if logfire_span is None:
+        return
+    span_data = get_stream_span_data(stream) | {"async": False}
+    logfire_span.set_attributes(span_data)
+
+
+def handle_base_model(
+    result: BaseModel | BaseStructuredStream, logfire_span: logfire.LogfireSpan | None
+):
+    if logfire_span is None:
+        return
+    span_data: dict[str, Any] = {}
+    output: dict[str, Any] = {}
+    if isinstance(result, BaseModel) and result._response is not None:  # type: ignore
+        response: BaseCallResponse = result._response  # type: ignore
+        span_data |= get_call_response_span_data(response)
+        output |= get_output(response)
+        output["response_model"] = result
+        span_data["output"] = output
+    elif isinstance(result, BaseStructuredStream):
+        span_data = get_stream_span_data(result.stream)
+        output["response_model"] = result.response_model
```

Comment:
`result.response_model` in this case is the type of the response model, not the actual response model output, which you can get as the final output of the structured stream.

Easiest way around this without having to iterate through would likely be to add a new field (something like `final_response_model` that you then update at the end of the stream, but this will still require at least waiting until the stream is exhausted since it'll be `None` otherwise (which I think you do looking at your iter replacement function in the middleware).

Reply:
what about `response_object`?

Reply:
Hmm, maybe `constructed_response_model`?

Reply:
sg

Reply:
done